source/slang/slang-ast-serialize.h

```diff
@@ -7,13 +7,188 @@
 #include "slang-ast-support-types.h"
 #include "slang-ast-all.h"
 
+#include "slang-ast-builder.h"
+
 #include "../core/slang-byte-encode-util.h"
 
 #include "../core/slang-stream.h"
 
 namespace Slang
 {
 
+/*
+AST Serialization Overview
+==========================
+
+The AST node types are generally types derived from the NodeBase. The C++ extractor is used to associate an ASTNodeType with
+every NodeBase type, such that casting is fast and simple and we have a simple integer to uniquely identify those types. The
+extractor also performs another task of associating with the type name all of the fields held in just that type. The definition
+of the fields is stored in an 'x macro' which is in the slang-ast-generated-macro.h file, for example
+
+```
+#define SLANG_FIELDS_ASTNode_DeclRefExpr(_x_, _param_)\
+    _x_(scope, (RefPtr<Scope>), _param_)\
+    _x_(declRef, (DeclRef<Decl>), _param_)\
+    _x_(name, (Name*), _param_)
+``
+
+For the type DeclRefExpr, this holds all of the fields held in just DeclRefExpr in this case `scope`, `declRef` and `name`.
+DeclRefExpr derives from Expr and this might hold other fields and so forth.
+
+The implementation makes a distinction between the 'native' types, the regular C++ in memory types and 'serial' types.
+Each serializable C++ type has an associated 'serial' type - with the distinction that it can be written out and (with perhaps some other data)
+read back in to recreate the C++ type. The serial type can be a C++ type, but is such it can be written and read from disk and still
+represent the same data. 
+
+We need a mechanism to be able to do do a conversion between native and serial types. To make the association we use the template
+
+```
+template <typename T>
+struct ASTSerialTypeInfo;
+```
+
+and specialize it for each native type. The specialization holds
+
+SerialType - The type that will be used to represent the native type
+NativeType - The native typs
+SerialAlignment - A value that holds what kind of alignment the SerialType needs to be serializable (it may be different from SLANG_ALIGN_OF(SerialType)!)
+toSerial - A function that with the help of ASTSerialWriter convert the NativeType into the SerialType
+toNative - A function that with the help of ASTSerialReader convert the SerialType into the NativeType
+
+It is useful to have a structure that holds the type information, so it can be stored. That is achieved with
+
+```
+template <typename T>
+struct ASTSerialGetType;
+```
+
+This template can be specialized for a specific native types - but all it holds is just a function getType, which returns a ASTSerialType*,
+which just holds the information held in the ASTSerialTypeInfo template, but additionally including the size of the SerialType.
```

Comment:
This is a difficult sentence to work through before finishing my morning coffee. I *think* what you are getting at is that you have a custom RTTI system for serialization, so that every type to be serialized is associated with some `ASTSerialType*`, and this template is the thing that handles fetching the RTTI for a given type `T`, while the `ASTSerialTypeInfo<T>` template is what is used to *define* the policy stuff or a given type.